frontend/apps/web/components/RunTImeline/RunTimeline.tsx

```diff
@@ -0,0 +1,457 @@
+import {
+  DropdownMenu,
+  DropdownMenuCheckboxItem,
+  DropdownMenuContent,
+  DropdownMenuTrigger,
+} from '@/components/ui/dropdown-menu';
+import {
+  Tooltip,
+  TooltipContent,
+  TooltipProvider,
+  TooltipTrigger,
+} from '@/components/ui/tooltip';
+import { cn } from '@/libs/utils';
+import { Timestamp } from '@bufbuild/protobuf';
+import { JobRunEvent, JobRunStatus } from '@neosync/sdk';
+
+import { JobRunStatus as JobRunStatusEnum } from '@neosync/sdk';
+import {
+  CheckCircledIcon,
+  CrossCircledIcon,
+  MinusCircledIcon,
+  MixerHorizontalIcon,
+} from '@radix-ui/react-icons';
+import {
+  addMilliseconds,
+  format,
+  formatDuration,
+  intervalToDuration,
+} from 'date-fns';
+import { ReactElement, useMemo, useState } from 'react';
+import Spinner from '../Spinner';
+import TruncatedText from '../TruncatedText';
+import { Badge } from '../ui/badge';
+import { Button } from '../ui/button';
+
+interface Props {
+  tasks: JobRunEvent[];
+  jobStatus?: JobRunStatusEnum;
+}
+
+export default function RunTimeline(props: Props): ReactElement {
+  const { tasks, jobStatus } = props;
+
+  // this should probably be better typed and using the ActivityStatus types
+  // but those aren't currenlty sent to the FE as a status but are represented in the
+  // type field
+  const [selectedStatuses, setSelectedStatuses] = useState<string[]>([
+    'running',
+    'completed',
+    'failed',
+    'canceled',
+  ]);
+
+  const formatFullDate = (date: Timestamp | Date | undefined) => {
+    if (!date) return 'N/A';
+
+    if (date instanceof Timestamp) {
+      return format(convertTimestampToDate(date), 'MM/dd/yyyy HH:mm:ss:SSS');
+    }
+
+    if (date instanceof Date) {
+      return format(date, 'MM/dd/yyyy HH:mm:ss:SSS');
+    }
+  };
+
+  const formatDate = (date: Date) => format(date, 'MM/dd/yyyy');
+  const formatTime = (date: Date) => format(date, 'HH:mm:ss:SSS');
+
+  const formatTaskDuration = (
```

Comment:
These should all exist outside of this function so they are rendered every time.
Always prefer static functions that live at the module level instead of being rendered every single time inside of the component.

There are a ton of functions in this main component that can be moved out as they are pure. It will make this component easier to understand.

Reply:
updated